wpilib/src/dio.rs

```diff
@@ -106,37 +93,19 @@ impl DigitalOutput {
     }
 
     /// Enable PWM for this output.
-    pub fn enable_pwm(&mut self, initial_duty_cycle: f64) -> HalResult<()> {
+    pub fn enable_pwm(self, initial_duty_cycle: f64) -> HalResult<DigitalPwm> {
```

Comment:
With this API, in the case of a failure, the DIO is lost. I'm not sure if its possible to recover here, or if it's worth trying, but if it is recoverable we could return `Result<DigitalPWM, (HalError,Self)> or something similar.

Reply:
Such a Result would make it difficult to Try.

Reply:
I agree, but in a realistic usecase where you want to recover from a failed upgrade, this API requires you to recreate the `DigitalOut`. Using `?` requires just `.map_err(|(e, _p)| e)`, which we can put in an example.

Reply:
I'm struggling to find a good reason to recover from an error in this scenario though. What would you do in such a scenario? The Nidec Brushless requires a DIO pin in PWM mode, for example.

Reply:
In the general case of a hardware output failure, you basically have two options:
1. Retry Later
2. Abandon and restart
3. Ignore it and carry on running your robot
The failure mode in which 1. could be useful is possible resource exhaustion `*status = NO_AVAILABLE_RESOURCES;` in `HAL_AllocateDigitalPWM`.

It's also more conducive to dependency injection. Say I get a `DigitalOutput` injected; I don't know it's port. How do I recover from this error? Maybe I just wanted to try using PWM if it's available. If I only want to use PWM sometimes, I'd like to keep my DigitalOutput around for other use.

These use cases / failure modes are unlikely enough that I'm willing to merge it either way, but that's my perspective.